### PR TITLE
Correct handling status of reparse point

### PR DESCRIPTION
**Pull request by stima**
Correct handling of placeholder file (https://msdn.microsoft.com/en-us/windows/compatibility/placeholder-files) in status. If 'is_reparse_point_a_symlink' returns false, return reparse_file as status. Otherwise resolve target and retake attributes of it.

### CODE REVIEW

**Comment by stima** [comment]
this is reopen of https://github.com/boostorg/filesystem/pull/74, I added test and did rebase.

**Comment by Lastique** [comment]
Please, rebase on the current develop, if this is still relevant.

**Comment by Lastique** [comment]
Also, please add a description of the problem.

**Comment by Lastique** [comment]
Please factor out the test to a separate file and condition it on the `check-mklink` rule. See [here](https://github.com/boostorg/filesystem/blob/fcbdae1581a8a334b7b5c1b9d6c15210a1451c44/test/Jamfile.v2#L80) for an example.

**Comment by Lastique** [comment]
Don't forget to add a license header comment and a short description of what the test verifies in the new test.

**Comment by Lastique** [comment]
Tests have [failed](https://ci.appveyor.com/project/Lastique/filesystem/builds/32660231/job/7fy5a199919v4066#L842).

File: test/issues/reparce_tag_file_placeholder.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+bool obtain_restore_privilege()
+{
+  HANDLE hToken;
+  if (!OpenProcessToken(GetCurrentProcess(), TOKEN_ADJUST_PRIVILEGES, &hToken))
+  {
+    std::cout << "OpenProcessToken() failed with: " << GetLastError() << std::endl;
+    return false;
+  }
+
+  BOOST_SCOPE_EXIT(&hToken)
```
Comment (Lastique) [inline test/issues/reparce_tag_file_placeholder.cpp:68]: Better not use the ScopeExit library. Just write a simple scope guard class.

File: test/issues/reparce_tag_file_placeholder.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+bool create_io_reparse_file_placeholder(const wchar_t* name)
+{
+  if (!obtain_restore_privilege())
+  {
+    return false;
+  }
+
+  HANDLE hHandle = CreateFileW(name, GENERIC_WRITE,
+    FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, CREATE_ALWAYS,
```
Comment (Lastique) [inline test/issues/reparce_tag_file_placeholder.cpp:100]: `#include <cstddef>` is missing for `NULL`.

File: test/issues/reparce_tag_file_placeholder.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    FILE_FLAG_OPEN_REPARSE_POINT, 0);
+
+  if (hHandle == INVALID_HANDLE_VALUE)
+  {
+    std::cout << "CreateFile() failed with: " << GetLastError() << std::endl;
+    return false;
+  }
+
+  auto pReparse = reinterpret_cast<PREPARSE_DATA_BUFFER>(GlobalAlloc(GPTR, MAXIMUM_REPARSE_DATA_BUFFER_SIZE));
```
Comment (Lastique) [inline test/issues/reparce_tag_file_placeholder.cpp:109]: `auto` is C++11. Spell the type completely.

**Comment by Lastique** [comment]
Still [failing](https://ci.appveyor.com/project/Lastique/filesystem/builds/32668507).

File: test/issues/reparce_tag_file_placeholder.cpp (hunk 1 above)
Comment (stima) [reply test/issues/reparce_tag_file_placeholder.cpp]: done

File: test/issues/reparce_tag_file_placeholder.cpp (hunk 2 above)
Comment (stima) [reply test/issues/reparce_tag_file_placeholder.cpp:98]: done

File: test/issues/reparce_tag_file_placeholder.cpp (hunk 3 above)
Comment (stima) [reply test/issues/reparce_tag_file_placeholder.cpp]: done

**Comment by Lastique** [comment]
Still [failing](https://ci.appveyor.com/project/Lastique/filesystem/builds/32680523/job/dxih671ij8sr558j).

**Comment by stima** [comment]
Thanks for helping to finalize it. There are only left:
error: "20" is not a known value of feature <cxxstd>
error: legal values: "98" "03" "0x" "11" "1y" "14" "1z" "17" "2a" "latest"

But this code does not affect gcc.

**Comment by Lastique** [comment]
Thanks. Don't mind the Boost.Build issue, it should be fixed by https://github.com/boostorg/build/pull/592.

**Pull request merged**